### PR TITLE
gh-156180: Reject a negative `len2` in `zlib.{adler32,crc32}_combine()`

### PR DESCRIPTION
**Pull request by StanFromIreland**
* Issue: gh-156180

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2026-08-21-14-30-00.gh-issue-156180.Qz3Lv8.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+:func:`zlib.crc32_combine` and :func:`zlib.adler32_combine` now raise
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2026-08-21-14-30-00.gh-issue-156180.Qz3Lv8.rst]: Very nitpicky: but can you put adler32 before crc32? I find it more natural to have the list alphabetically sorted.+

File: Modules/zlibmodule.c [hunk 2]
```diff
@@ -1951,6 +1951,10 @@ zlib_adler32_combine_impl(PyObject *module, unsigned int adler1,
     if (PyErr_Occurred()) {
         return (unsigned int)-1;
     }
+    if (len < 0) {
+        PyErr_SetString(PyExc_ValueError, "len2 must be non-negative");
+        return (unsigned int)-1;
+    }
```
Comment (picnixz) [inline Modules/zlibmodule.c:1956]:
```suggestion
    if (len < 0) {
        if (!PyErr_Occurred()) {
            PyErr_SetString(PyExc_ValueError, "len2 must be non-negative");
        }
        return (unsigned int)-1;
    }
```

Comment (picnixz) [reply Modules/zlibmodule.c:1956]: And likewise for the crc32_combine

File: Lib/test/test_zlib.py [hunk 3]
```diff
@@ -182,6 +182,7 @@ def test_combine_no_iv_invalid_length(self):
             self.assertNotEqual(invalid_res, checksum)
 
         self.assertRaises(TypeError, self.combine, 0, 0, "len")
+        self.assertRaises(ValueError, self.combine, 0, 0, -1)
```
Comment (picnixz) [inline Lib/test/test_zlib.py:185]: Can you add a test for overflow errors as well? I think I forgot (or maybe there is no such case and we still have a ValueError)

File: Misc/NEWS.d/next/Library/2026-08-21-14-30-00.gh-issue-156180.Qz3Lv8.rst (hunk 1 above)
Comment (StanFromIreland) [reply Misc/NEWS.d/next/Library/2026-08-21-14-30-00.gh-issue-156180.Qz3Lv8.rst]: Really can't get much nitier, but sure ;-)

File: Lib/test/test_zlib.py (hunk 3 above)
Comment (serhiy-storchaka) [inline Lib/test/test_zlib.py:185]: While we are here, it would be nice to tests for `2**1000` and `-2**1000`.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM. 👍

File: Modules/zlibmodule.c [hunk 4]
```diff
@@ -1948,7 +1948,10 @@ zlib_adler32_combine_impl(PyObject *module, unsigned int adler1,
 #else
     z_off_t len = convert_to_z_off_t(len2);
 #endif
-    if (PyErr_Occurred()) {
+    if (len < 0) {
```
Comment (picnixz) [inline Modules/zlibmodule.c:1951]: For a follow-up, we should have `convert_to_z_off_t` also reject negative integers if given a flag ("allow negative" for instance). Also the logic of combine() is actually duplicated for the length. I guess we could have a single function for that (I don't remember why I didn't do so at that time).

**Review by picnixz — APPROVED** [review]

**Comment by StanFromIreland** [comment]
Thanks for the reviews!
